go/lang/security/injection/tainted-url-host.yaml

```diff
@@ -53,4 +40,58 @@ rules:
     - metavariable-regex:
         metavariable: $ANYTHING
         regex: ^(BasicAuth|Body|Cookie|Cookies|Form|FormValue|GetBody|Host|MultipartReader|ParseForm|ParseMultipartForm|PostForm|PostFormValue|Referer|RequestURI|Trailer|TransferEncoding|UserAgent|URL)$
+  - label: FORMAT
+    requires: INPUT
+    patterns:
+    - pattern-either:
+      - patterns:
+        - pattern-either:
+          - pattern: fmt.Fprintf($F, "$URLSTR", $INPUT, ...)
+          - pattern: fmt.Sprintf("$URLSTR", $INPUT, ...)
+          - pattern: fmt.Printf("$URLSTR", $INPUT, ...)
+        - metavariable-regex:
+            metavariable: $URLSTR
+            regex: ^((http://)|(https://)|(//))%(v|s|q).*
+    - focus-metavariable: $INPUT
+  - label: ADD
+    requires: INPUT
+    patterns:
+    - pattern-either:
+      - pattern: |
+          $F + $X
+      - pattern: fmt.Fprintf(..., $X, ...)
+      - pattern: fmt.Sprintf(..., $X, ...)
+      - pattern: fmt.Printf(..., $X ...)
+    - pattern-not: |
+        "https://" + $X
+    - pattern-not: |
+        "http://" + $X
+    - pattern-not: |
+        "//" + $X
+    - pattern-not: |
+        $X + $F
+    - focus-metavariable: $X
+  pattern-sinks:
+  - requires: FORMAT or (not ADD)
```

Comment:
Should this be `FORMAT and (not ADD)` ?

Reply:
good question, it should be "or" because we are looking for something that has string concat via format strings with something looking like `http/https` OR something that is NOT `http://www.example.com/` + `<user input>` (doesn't matter bc its only path that is being controlled). 

It's essentially two different options of concat that we're looking at: via format string and via `+` operator, and we want either option to work.

Reply:
additional clarification - does this negation construct for labels behave intuitively? 

consider:

```go
  err_redirect := "https://app.foo.com/error"
  // stuff happens
  if err != nil {
    http.Redirect(w, r, err_redirect, http.StatusSeeOther)
  } 
```

I'm assuming the redirect string doesn't get tainted because it's not a source (doesn't match `INPUT` requirements). So, things with the `INPUT` label (and nothing else) can make it into the sink as long as they don't get the `ADD` label?

This *feels* like a sanitizer construction but I understand the use of the labeling here, at least intuitively.  Is there an equivalent sanitizer construction? Is it more performant to use sanitizers vs this sort of label conditional?

Reply:
These are very good points that ya'll have brought up -- @kurt-r2c I will look into trying this with sanitizers today, and @IagoAbal thank you for clarification with the taint labels. I'll try messing around with `FORMAT or (INPUT and not ADD)`. Thanks for fixing the core issue!